src/lib/backoff.ts

```diff
@@ -0,0 +1,66 @@
+import { promisify } from 'util';
+
+const sleep = promisify(setTimeout);
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (msg: string) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: 10,
+};
+
+/**
+ * Retries an async function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options to allow backoff caller to log provided message.
+ */
+export default async function backoff(
+	fn: any,
+	options: Partial<Options> = {},
+): Promise<any> {
+	if (typeof fn !== 'function') {
+		throw new Error('Must pass a function to backoff');
+	}
+
+	try {
+		return await fn();
+	} catch (e) {
+		if (e.retryAfter) {
+			options.attempt = 0;
+			options.minDelay = e.retryAfter;
+		}
+	}
+
+	const config: Options = {
+		...DEFAULT_OPTIONS,
+		...options,
+	} as Options;
+
+	const wait = delay(config.attempt, config.minDelay, config.maxDelay);
+
+	if (config.retryCallback) {
+		config.retryCallback(`Retrying request in ${wait / 1000} seconds`);
+	}
+
+	await sleep(wait);
+
+	if (config.maxRetries && config.attempt + 1 >= config.maxRetries) {
+		throw new Error(`Reached max number of retries: ${config.maxRetries}`);
+	}
```

Comment:
why throw error of retries reached after waiting ? move this to before retryCallback since the function is not going to retry.